torch/csrc/jit/codegen/cuda/lower_loops.cpp

```diff
@@ -282,46 +233,123 @@ void LoopNestGenerator::handle(Expr* expr) {
   }
 
   TensorView* out = expr->output(0)->as<TensorView>();
-  // 1) Reduce loop structure
-  while (compute_at_scope.size() > out->getThisComputeAtAxis() &&
-         compute_at_scope.back().second != out &&
-         compute_at_scope.back() !=
-             out->getComputeAtAxis((int)compute_at_scope.size() - 1)) {
-    popFor();
+
+  // Figure out what the entire loop structure should look like.
+  std::deque<std::pair<IterDomain*, TensorView*>> loop_structure;
+
+  // As we go through iteration domains track the previous view
+  TensorView* last_ca_view = nullptr;
+  // Check where in the previous view our last axis was in that view
+  int64_t last_ca_view_ind = 0;
+
+  // Look at each axis individually in out's domain
+  for (int64_t out_i = 0; out_i < (int64_t)out->getThisComputeAtAxis();
+       out_i++) {
+    // Grab the axis information
+    auto ca_point = out->getComputeAtAxis(out_i);
+    auto ca_view = ca_point.second;
+    auto ca_id = ca_point.first;
+
+    // Figure out if there are axes in the compute at tensor view that aren't
+    // in out, make sure to also open them. Check where to start looking for
+    // them in the compute at view.
+    size_t start = 0;
+    if (last_ca_view == nullptr) {
+      // Start at the begining, we haven't processed any axes yet.
+      start = 0;
+    } else if (last_ca_view == ca_view) {
+      // This view is the same as the last axis, so start where we left off.
+      start = last_ca_view_ind + 1;
+    } else {
+      // This is a new view, figure out where we are in it, and start from there
+      for (start = 0; start < ca_view->nDims(); start++) {
+        if (loop_structure.back().first ==
+            ca_view->getComputeAtAxis(start).first) {
+          break;
+        }
+      }
+      start++;
+    }
+
+    // Go from start, and open all loops in the computeAt view until we hit the
+    // one associated with out->getComputeAtAxis(out_i)
+    for (size_t ca_i = start; ca_i < ca_view->nDims(); ca_i++) {
+      loop_structure.push_back(ca_view->getComputeAtAxis(ca_i));
```

Comment:
Is it true that, here, `ca_view->getComputeAtAxis(ca_i).second == ca_view` and `ca_view->getComputeAtAxis(ca_i).first== ca_view->axis(ca_i)`? These tests are at least the case for the test_gpu tests.

Reply:
I believe you are correct.